### DISCUSSION TITLE
Clarify server pool and erasure set availability in some scenarios

### DISCUSSION THREAD

**Discussion by omarkohl** [minio/minio#16870] [Q&A]
> MinIO can tolerate the loss of up to N drives per erasure set and continue performing read and write operations (“quorum”). If N is equal to exactly 1/2 the drives in the erasure set, MinIO write quorum requires drives to avoid data inconsistency (“split-brain”).

https://min.io/docs/minio/linux/operations/concepts/erasure-coding.html

Based on the above quote, your clarification concerning server pool behavior in https://github.com/minio/minio/discussions/16840 and the following hypothetical scenario, could you please clarify how the availability is affected?

Scenario:

* Server pool 1
  * 4 nodes (servers) with 8 disks each
  * Erasure set size 16 (i.e. 4 disks per node)
  * Parity 2
* Server pool 2 (same configuration)
  * 4 nodes with 8 disks each
  * Erasure set size 16 (i.e. 4 disks per node)
  * Parity 2

Here a graphical representation:

![minio-ec_concepts-server-pool-scale-out](https://user-images.githubusercontent.com/1420656/226853004-4f3ca19c-1072-41bd-82a2-6dc436df7ebb.png)

How would the availability (READ and WRITE) of the cluster be affected:

* If in the erasure set on the upper left 2 disks fail? I assume the entire cluster would still do READ and WRITE correctly.
* If in the erasure set on the upper left 3 disks fail? Would the entire cluster (both pools) still do READs? Would all WRITEs be redirected to pool 2? When trying to READ an object stored in the affected erasure set, would that operation fail?
* If for both erasure sets on the left, 3 disks fail in each one? Does that count as pool 1 being down, making the entire cluster go down?

**Comment by ravindk89** [comment]
Since your parity is 2, if you lose 2 drives, you are holding quorum - but just by your fingernails.

Now my understanding of drive allocation is we kind of choose one per node going down the list of nodes, then circle back. So something like...

- Node 1 Disk 1 - erasure set 1
- Node 2 Disk 1 - erasure set 1
- Node 3 Disk 1 - erasure set 1
- Node 4 Disk 1 - erasure set 1
- Node 1 Disk 2 - erasure set 1 ...

So if you lost two disks on, say, node 1 - there's a chance you've lost one disk per erasure set and actually might be OK for a little while longer. *But* that's theoretical, you'd still want to resolve those disk replacements ASAP considering the very low parity. Having a larger number of nodes per pool kind of helps a bit here.

As far as losing *quorum* - based on internal discussions, we have to halt IO for the cluster only when the entire pool goes down*. For individual erasure sets, MinIO makes best-efforts to route PUT/GET/DELETE, but if the object hashes to a degraded erasure set, you should get an error.

Once you lose quorum, losing more disks is kind of just adding fuel to the fire. Losing quorum means data loss. So if you lose two erasure sets worth of quorum, you've lost two erasure sets worth of data. The rest of your data on the second pool is safe on disk, at minimum. At that point we have to hope you have Replication set up (site or bucket) where you could restore the lost data.

This is partially why we don't recommend EC < 4 - you're risking a whole lot of sadness if you have a bad batch of drives.

I'm going to be revisiting our docs on erasure sets in the near future to try to make this a little bit cleaner.

EDITS made pending internal discussion.

**Reply by omarkohl** [reply]
If I understand you correctly, once you lose quorum within a single erasure set the IO for the complete cluster will stop. Is that correct? So in my image above, if 3 disks within the top left erasure set (light blue box) are lost then every S3 PUT/GET directed at any of the nodes would fail. Is that correct?

What about the case when using parity 8 and erasure set size of 16 (i.e. parity is half the disks of the erasure set) and losing 8 disks within a single erasure set, would that lead to READ-only access for the entire cluster? Because for that erasure set, WRITE will no longer be possible. See the below quote:

> MinIO can tolerate the loss of up to N drives per erasure set and continue performing read and write operations (“quorum”). If N is equal to exactly 1/2 the drives in the erasure set, MinIO write quorum requires N+1 drives to avoid data inconsistency (“split-brain”).

Source: https://min.io/docs/minio/linux/operations/concepts/erasure-coding.html

> Now my understanding of drive allocation is we kind of choose one per node going down the list of nodes, then circle back.

That is also my understanding. My diagram is meant in that way, so the disks would be assigned to erasure sets as depicted.

```
Node1
-------------------
Disk1      Disk2
Disk3      Disk4
Disk5      Disk6
Disk7      Disk8
-------------------
```

> So if you lost two disks on, say, node 1 - there's a chance you've lost one disk per erasure set and actually might be OK for a little while longer.

I am aware of that. I am specifically asking what would happen in the case of losing quorum within erasure sets. I understand that it's an unlikely scenario.

> This is partially why we don't recommend EC < 4 - you're risking a whole lot of sadness if you have a bad batch of drives.

My example is hypothetical to more easily understand the behavior of MinIO. I am aware that relying solely on EC:2 without replication is a bad idea. Nevertheless, losing quorum is always a possibility, no matter how much parity you have, even though of course it gets more and more unlikely.

**Reply by klauspost** [reply]
There is a difference between *lock quorum* and *object quorum*. If you lose quorum on a set (which covers a number of objects, but not all), you will be unable to read/write depending on how much you lose. These objects will be unavailable, but the rest remain available.

If you lose *lock quorum* you will be unable to perform operations on the objects as well, but the thresholds are are different. If a cluster looses lock quorum it will not be able to operate, since it will not be able to acquire locks for objects.

These concepts are linked, but not the same.

> If I understand you correctly, once you lose quorum within a single erasure set the IO for the complete cluster will stop. Is that correct?

No. Some operations like listings will be unable to complete, since it requires merging all sets, but reads and writes operate independently on sets and can proceed if the set the object is (or would be) placed in is available. It also depends on whether disks are offline, but the host is reachable, and more factors.

But practically operations will be severely limited if a set isn't operational.

**Reply by omarkohl** [reply]
So if we assume my example where the upper left erasure set (let's call it Set1) loses 3 disks but all nodes remain online, are the following statements true?

* It is possible to READ any object, except those located on Set1?
* It is possible to WRITE any object that would be placed in Pool2?

Assignment to pools is non-deterministic, so what would happen if the object _would_ end up on Pool1? Would it then succeed if within Pool1 it would be written to Set2 but fail if within Pool1 it would be written to Set1?

**Reply by klauspost** [reply]
Before writes each pool is queried if the object exists there, so objects don't exist in multiple pools. 

If pool 1 is unavailable, the write will fail regardless AFAIR.

**Reply by AlexZIX** [reply]
I still can't understand how minio distribute disks between erasure sets in distributed setup. Say we have 4 nodes with 20 disks on each (total 80 disks). Erasure set size in this case will be 16. We choose the EC = 8. So we can lost up to 40 disks or 2 nodes according minio calculator (https://min.io/product/erasure-code-calculator). But if 9 disks (EC+1) of one erasure set be located at same node we will lost data in case of 1 node becomes unavailable (or even half of this node when 10 disks located in first disk enclosure connected to this node and 10 disks in second enclosure for example).
So after some thinking I created following rule for me to prevent unintended behavior: total disks count per pool shouldn't be above 16. In this case I'll be sure that one erasure set covers all disks in pool. If I need to use more than 16 disks I just pool it using ZFS. For now I have following minio setup in my (small) production: 4 nodes with 24 disks on each (96 SAS 600 GB disks total). On each node 24 disks separated into 4 pools of 6 disks using ZFS. EC = 6. So MinIO see the 16 disks and creates erasure set with stripe size 16. I can lost 6 disks polls which is 36 physical disks. When I need to expand the storage I can just extend the disks pools - MinIO still see 16 disks total. Of course it's not a good idea to expand this setup infinitely using this way because of long healing after disk failure. So after the each disks pool becomes 3 TB I just add new MinIO pool with ZFS pools.
As the additional benefit - ZFS can monitor its pools and disable one of them in case of degradation one of disks unlike the MinIO which degrades entire cluster while you are looking for a problem.

**Reply by kannappanr** [reply]
@AlexZIX In your case, MinIO will allocate only 4 disks per node within an erasure set

**Reply by AlexZIX** [reply]
@kannappanr Yes, it's the why I use ZFS for disks pooling. Physically I have 96 disks which pools into 16 groups of 6 disks. So Minio see the 4 disks (pools) on each node.

root@minio-cold-1:~# df -h
Filesystem                         Size  Used Avail Use% Mounted on
hdd-pool-4                         3.2T  1.1T  2.1T  34% /hdd-pools/hdd-pool-4
hdd-pool-3                         3.2T  1.1T  2.1T  34% /hdd-pools/hdd-pool-3
hdd-pool-1                         3.2T  1.1T  2.1T  34% /hdd-pools/hdd-pool-1
hdd-pool-2                         3.2T  1.1T  2.1T  34% /hdd-pools/hdd-pool-2

root@minio-cold-1:~# zpool list
NAME         SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
hdd-pool-1  3.26T  1.07T  2.19T        -         -     9%    32%  1.00x    ONLINE  -
hdd-pool-2  3.26T  1.07T  2.19T        -         -    10%    32%  1.00x    ONLINE  -
hdd-pool-3  3.26T  1.07T  2.19T        -         -     9%    32%  1.00x    ONLINE  -
hdd-pool-4  3.26T  1.07T  2.19T        -         -     9%    32%  1.00x    ONLINE  -

root@minio-cold-1:~# zpool status
  pool: hdd-pool-1
 state: ONLINE
  scan: scrub repaired 0B in 01:30:34 with 0 errors on Sun Mar 12 01:54:35 2023
config:

        NAME                                       STATE     READ WRITE CKSUM
        hdd-pool-1                                 ONLINE       0     0     0
          scsi-SHITACHI_HUC10606_CLAR600_PPK6M9XB  ONLINE       0     0     0
          scsi-SHITACHI_HUC10606_CLAR600_PPK71BDB  ONLINE       0     0     0
          scsi-SHITACHI_HUC10606_CLAR600_PPKTEWBB  ONLINE       0     0     0
          scsi-SHITACHI_HUC10606_CLAR600_PPKW223B  ONLINE       0     0     0
          scsi-SHITACHI_HUC10606_CLAR600_PPKY03ZB  ONLINE       0     0     0
          scsi-SHITACHI_HUC10606_CLAR600_PPKY1EPB  ONLINE       0     0     0

errors: No known data errors

**Reply by AlexZIX** [reply]
Additionally I can simple see the current load and iops:

root@minio-cold-1:~# zpool iostat
              capacity     operations     bandwidth
pool        alloc   free   read  write   read  write
----------  -----  -----  -----  -----  -----  -----
hdd-pool-1  1.07T  2.19T      7     15   368K   455K
hdd-pool-2  1.07T  2.19T      7     15   374K   450K
hdd-pool-3  1.07T  2.19T      6     15   372K   447K
hdd-pool-4  1.07T  2.19T      6     15   370K   448K
----------  -----  -----  -----  -----  -----  -----
